### PR TITLE
[FSDP()][14/N] Refactor pre-forward/post-backward

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87937 [FSDP] Simplify `_reset_lazy_init()`
* #87936 [FSDP()][22/N] Refactor `_cast_buffers()` in `_lazy_init()`
* #87935 [FSDP()][21/N] Refactor `_buffer_name_to_orig_dtype` computation
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* **#87927 [FSDP()][14/N] Refactor pre-forward/post-backward**
* #87926 [FSDP()][13/N] Refactor unshard/reshard/grads
* #87925 [FSDP()][12/N] Easy cleanup
* #87924 [FSDP()][10/N][11/N] Introduce composable (ctor only)
* #87923 [FSDP()][9/N] Refactor ctor (continued)

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    p_assert(len(handles_key) > 0, "Expects a non-empty handles key")
+    training_states = set(handle._training_state for handle in handles_key)
+    p_assert(
+        len(training_states) == 1,
+        f"Expects uniform training state but got {training_states}",
+    )
+    return next(iter(training_states))
+
+
+def _register_post_backward_hooks(
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:440]: Simple move

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    state: _State,
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:441]: Note to self: `state` seems unused.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+            for target_handles_key in eod.get_handles_to_forward_prefetch(
+                current_handles_key
+            )
+            if state._needs_pre_forward_unshard.get(target_handles_key, False)
+            and not state._handles_prefetched.get(target_handles_key, False)
+        ]
+    return target_handles_keys
+
+
+def _get_training_state(
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:427]: Simple move

File: torch/distributed/fsdp/_runtime_utils.py [hunk 4]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    for handles_key in handles_to_prefetch:
+        # Prefetch the next set of handles without synchronizing to allow
+        # the sync to happen as late as possible to maximize overlap
+        _unshard(
+            state, handles_key, state._streams["unshard"], state._streams["pre_unshard"]
+        )
+        state._handles_prefetched[handles_key] = True
+
+
+@no_type_check
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:374]: Simple move

File: torch/distributed/fsdp/_runtime_utils.py [hunk 5]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    )
+
+
+@no_type_check
+def _low_precision_hook_enabled(state: _State) -> bool:
+    return state._communication_hook in LOW_PRECISION_HOOKS
+
+
+@no_type_check
+def _prefetch_handles(
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:354]: Simple move

File: torch/distributed/fsdp/_runtime_utils.py [hunk 6]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    reduced precision for communication.
+    However, if a low precision communication hook is registered, then this
+    dtype cast happens in the hook instead.
+    """
+    _assert_in_training_states(state, [TrainingState.FORWARD_BACKWARD])
+    if not _low_precision_hook_enabled(state) and (
+        handle._uses_param_mixed_precision or handle._uses_reduce_mixed_precision
+    ):
+        low_prec_grad_data = sharded_grad.data
+        sharded_grad.data = sharded_grad.data.to(dtype=param.dtype)
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:311]: Added comment to clarify `record_stream()`

File: torch/distributed/fsdp/_runtime_utils.py [hunk 7]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+            if handle._use_orig_params:
+                # Since the handle's `FlatParameter` completed its gradient
+                # computation, we should reset the gradient noneness mask
+                handle._reset_is_grad_none()
+                # Delay using sharded gradient views until after the
+                # reduce-scatter instead of immediately after resharding
+                handle._use_sharded_grad_views()
+
+
+@no_type_check
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:275]: Renamed

File: torch/distributed/fsdp/_runtime_utils.py [hunk 8]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
         handle.reshard_grad()
 
 
+@no_type_check
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:109]: Simple move

File: torch/distributed/fsdp/_runtime_utils.py [hunk 9]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+    """
+    state.training_state = TrainingState.FORWARD_BACKWARD
+    state._exec_order_data.record_pre_forward(handles, module.training)
+    for handle in handles:
+        handle._training_state = HandleTrainingState.FORWARD
+    if unshard_fn is not None:
+        unshard_fn()
+    # Register post-backward hooks to reshard the parameters and reduce-scatter
+    # their gradients. They must be re-registered every forward pass in case
+    # the `grad_fn` is mutated.
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:138]: would I be correct, we can use module backward hook and avoid registering every iteration?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 10]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+        p_assert(
+            temp_flat_param.grad_fn is not None,
+            "The `grad_fn` is needed to access the `AccumulateGrad` and "
+            "register the post-backward hook",
+        )
+        acc_grad = temp_flat_param.grad_fn.next_functions[0][0]
+        hook_handle = acc_grad.register_hook(
+            functools.partial(_post_backward_hook, state, handle)
+        )
+        flat_param._post_backward_hook_state = (acc_grad, hook_handle)  # type: ignore[attr-defined]
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:476]: do we need to worry about circular reference here?

handle -> flat_param -> _post_backward_hook_state -> acc_grad -> handle

File: torch/distributed/fsdp/_runtime_utils.py [hunk 11]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+                param.grad.data = param.grad.to(state.mixed_precision.reduce_dtype)
+
+            if handle.uses_sharded_strategy:
+                # We clear `.grad` to permit multiple backwards. This avoids a
+                # race where the second backward pass computation precedes
+                # ahead of the first backward pass reduction, which is possible
+                # since the reduction is issued in a separate stream and is
+                # async and would result in reducing the wrong gradient.
+                unsharded_grad = param.grad.data
+                param.grad = None
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:214]: Does this mean `no_sync()` won't work?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 12]
```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
...
+                p_assert(
+                    len(unsharded_grad.size()) == 1,
+                    f"Expects gradient to be flattened but got {unsharded_grad.size()}",
+                )
+                chunks = list(unsharded_grad.chunk(state.world_size))
+                numel_to_pad = (
+                    state.world_size * chunks[0].numel() - unsharded_grad.numel()
+                )
+                padded_unsharded_grad = F.pad(unsharded_grad, [0, numel_to_pad])
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:223]: will this create a new and larger padded tensor for unsharded grad? If yes, it seems more reasonable to keep the padded one and `param.grad` points to its view?

**Review by mrshenli — APPROVED** [review]

**Pull request closed without merging**